pkg/runtime/grpc/k8s-exec-dialer.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcruntime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/factory"
+)
+
+type k8sExecConn struct {
+	io.Writer
+	io.Reader
+	exec       remotecommand.Executor
+	podName    string
+	cancelFunc func()
+}
```

Comment:
We still have to do the exec (however, we do it programmatically like we did before and not by calling `kubectl exec`). This just adds a wrapper around that exec so it's handled just like a network connection, instead of marshalling/unmarshalling/parsing lines of output. We still have to use `exec`, as port forwarding isn't available for unxi sockets.

Reply:
> Hum... We rather do a `kubectl exec` over the k8s REST API.
> I then think I am missing something...
> At a first glance, I thought using `gRPC` will avoid us doing a `kubectl exec`.
> Indeed here, the `kubectl exec` is no more done from CLI. But in terms of scaling, what is the impact of this?
> Also, can't we avoid doing this `kubectl exec`?

In terms of scaling, this is (still) not optimal. It still uses the kubernetes API server (so all traffic goes through it) and even if we would use portforwarding (to a tcp rather than a unix socket) instead of exec, the node would do essentially the same and spawn a process (I think it's socat as well) to tunnel the traffic. The high throughput solution with better scaling would be to expose the gadget service using tcp and directly talking to it. I've done that in #1096, but we decided to postpone it and clean up for now. Using the changes in this PR however, it's fairly easy to re-add that on top using the same gadget API.